### PR TITLE
Implement some more environment functions for WASI.

### PR DESCRIPTION
**Pull request by brendandburns**
Fixes https://github.com/ziglang/zig/issues/14135

Due to the limitations of the functionality that WASI currently exposes, this isn't great from a performance perspective, but it probably doesn't matter that much and it's definitely better than just returning EnvVar not found.

### CODE REVIEW

**Comment by brendandburns** [comment]
I don't know what to make of the failures, the error message is this:

```
[5/21] Linking C executable zig-wasm2c
In file included from /home/ci/deps/zig+llvm+lld+clang-aarch64-linux-musl-0.11.0-dev.971+19056cb68/lib/zig/libc/musl/crt/rcrt1.c:3:
/home/ci/deps/zig+llvm+lld+clang-aarch64-linux-musl-0.11.0-dev.971+19056cb68/lib/zig/libc/musl/crt/../ldso/dlstart.c:146:20: warning: a function declaration without a prototype is deprecated in all versions of C and is treated as a zero-parameter prototype in C2x, conflicting with a subsequent definition [-Wdeprecated-non-prototype]
        GETFUNCSYM(&dls2, __dls2, base+dyn[DT_PLTGOT]);
                          ^
/home/ci/deps/zig+llvm+lld+clang-aarch64-linux-musl-0.11.0-dev.971+19056cb68/lib/zig/libc/musl/crt/rcrt1.c:11:13: note: conflicting prototype is here
hidden void __dls2(unsigned char *base, size_t *sp)
            ^
1 warning generated.
```

**Comment by brendandburns** [comment]
oops, I didn't mean to close this, pushed the wrong button...

**Comment by Vexu** [comment]
That warning is unrelated, the actual error is:
```
[15/21] Running zig1.wasm to produce /home/ci/actions-runner4/_work/zig/zig/build-new/compiler_rt.c
FAILED: compiler_rt.c 
cd /home/ci/actions-runner4/_work/zig/zig && /home/ci/actions-runner4/_work/zig/zig/build-new/zig1 /home/ci/actions-runner4/_work/zig/zig/lib build-obj lib/compiler_rt.zig -ofmt=c -OReleaseSmall --name compiler_rt -femit-bin="/home/ci/actions-runner4/_work/zig/zig/build-new/compiler_rt.c" --pkg-begin build_options /home/ci/actions-runner4/_work/zig/zig/build-new/config.zig --pkg-end -target x86_64-linux
unimplemented
Aborted
[16/21] Running zig1.wasm to produce /home/ci/actions-runner4/_work/zig/zig/build-new/zig2.c
FAILED: zig2.c 
cd /home/ci/actions-runner4/_work/zig/zig && /home/ci/actions-runner4/_work/zig/zig/build-new/zig1 /home/ci/actions-runner4/_work/zig/zig/lib build-exe src/main.zig -ofmt=c -lc -OReleaseSmall --name zig2 -femit-bin="/home/ci/actions-runner4/_work/zig/zig/build-new/zig2.c" --pkg-begin build_options /home/ci/actions-runner4/_work/zig/zig/build-new/config.zig --pkg-end -target x86_64-linux
unimplemented
Aborted
ninja: build stopped: subcommand failed.
Error: Process completed with exit code 1.
```

Likely coming form one of these: https://github.com/ziglang/zig/blob/af197d495478f7a5c33a6550d5525e86d797dc89/stage1/wasi.c#L658-L678

**Comment by brendandburns** [comment]
@Vexu 

Ok, so I think that the problem here is that Zig itself makes a few calls to `getEnvVarOwned` presumably in the `zig1` executable and since those are marked as 'unimplemented' in `wasi.c` this is now failing. Previous to this, when there wasn't a wasi-specific implementation of `getEnvVarOwned` they just returned empty strings.


There are a few calls that I can find that I think might be causing this are

https://github.com/ziglang/zig/blob/4172c2916684655a9742de99384be8110922ed07/src/main.zig#L631

https://github.com/ziglang/zig/blob/4172c2916684655a9742de99384be8110922ed07/src/introspect.zig#L86

https://github.com/ziglang/zig/blob/4172c2916684655a9742de99384be8110922ed07/lib/std/zig/system/NativePaths.zig#L30

I'm wondering what the right thing to do here is. My best guess is that since the calls to `getEnvVarOwned` were previously just returning empty strings, the right thing to do is to just remove the panic in `wasi.c` and instead return empty variables, since that will fix this issue and matches the previous behavior. (and I suspect those environment variables aren't desirable in the `zig1` stage anyway)

However, it's possible that there's a better solution, let me know what you think.

**Comment by Vexu** [comment]
I don't think the panic should be removed, either the function should be implemented properly or it should not be called which can be done by checking for the `only_c` build option which is set to true when building the wasi bootstrap binary (look for `build_options.only_c`).

**Comment by brendandburns** [comment]
ok, sounds good. Just so I have a sense for what `wasi.c` is doing. Is this basically a native-c implementation that kind of pretends to be the WASI systemcalls, but actually just uses the underlying operating system?

(that's what it looks like to me)

I'm happy to implement those functions in `wasi.c` if it is useful.

**Comment by Vexu** [comment]
The wasi bootstrap binary doesn't use environment variables so implementing them wouldn't be too useful. I guess making them no-ops would also be valid as long as it is clearly mentioned.

File: lib/std/process.zig [hunk 1]
```diff
@@ -369,6 +369,13 @@ pub fn getEnvVarOwned(allocator: Allocator, key: []const u8) GetEnvVarOwnedError
             error.UnexpectedSecondSurrogateHalf => return error.InvalidUtf8,
             else => |e| return e,
         };
+    } else if (builtin.os.tag == .wasi and !builtin.link_libc) {
+        var envmap = getEnvMap(allocator) catch return error.OutOfMemory;
+        defer envmap.deinit();
+        const val = envmap.getPtr(key) orelse return error.EnvironmentVariableNotFound;
+        const result = try allocator.alloc(u8, val.len);
+        mem.copy(u8, result, val.*);
+        return result;
```
Comment (Vexu) [inline lib/std/process.zig:378]:
```suggestion
        const val = envmap.get(key) orelse return error.EnvironmentVariableNotFound;
        return allocator.dupe(u8, val);
```

**Comment by brendandburns** [comment]
@Vexu thanks for the feedback, I'm still learning idiomatic Zig + standard library capabilities. Added your suggestion to the commits.

**Comment by Vexu** [comment]
Sorry for not noticing during previous reviews but you should also add a `@compileError` for using `hasEnvVarConstant` on wasi without libc since that would still trigger the issue you reported.

**Comment by brendandburns** [comment]
@Vexu comment addressed (I hope :)

**Comment by brendandburns** [comment]
Looks like there are some callers of that code in stage3 I will clean it up...

**Comment by matu3ba** [comment]
Looks like `stage3-debug/bin/zig build update-zig1` fails now:

```
2023-01-04T07:47:49.3020845Z + stage3-debug/bin/zig build update-zig1
2023-01-04T07:48:08.8642826Z stage3-debug/lib/zig/std/process.zig:388:9: error: hasEnvVarConstant is not supported for WASI without libc
2023-01-04T07:48:08.8644067Z         @compileError("hasEnvVarConstant is not supported for WASI without libc");
2023-01-04T07:48:08.8644572Z         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2023-01-04T07:48:08.8644937Z referenced by:
2023-01-04T07:48:08.8645588Z     buildOutputType: /home/ci/actions-runner7/_work/zig/zig/src/main.zig:675:35
2023-01-04T07:48:08.8646224Z     main: /home/ci/actions-runner7/_work/zig/zig/src/main.zig:195:20
2023-01-04T07:48:08.8647059Z     remaining reference traces hidden; use '-freference-trace' to see all reference traces
2023-01-04T07:48:08.8647325Z 
2023-01-04T07:48:08.8647562Z stage3-debug/lib/zig/std/process.zig:388:9: error: hasEnvVarConstant is not supported for WASI without libc
2023-01-04T07:48:08.8647853Z         @compileError("hasEnvVarConstant is not supported for WASI without libc");
2023-01-04T07:48:08.8648087Z         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2023-01-04T07:48:08.8648420Z stage3-debug/lib/zig/std/process.zig:388:9: error: hasEnvVarConstant is not supported for WASI without libc
2023-01-04T07:48:08.8648734Z         @compileError("hasEnvVarConstant is not supported for WASI without libc");
2023-01-04T07:48:08.8648946Z         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2023-01-04T07:48:08.8975915Z error: zig...
2023-01-04T07:48:08.8976348Z error: The following command exited with error code 1:
2023-01-04T07:48:08.8978661Z /home/ci/actions-runner7/_work/zig/zig/build-debug/stage3-debug/bin/zig build-exe --stack 33554432 /home/ci/actions-runner7/_work/zig/zig/src/main.zig -OReleaseSmall --cache-dir /home/ci/actions-runner7/_work/zig/zig/build-debug/zig-local-cache --global-cache-dir /home/ci/actions-runner7/_work/zig/zig/build-debug/zig-global-cache --name zig -target wasm32-wasi -mcpu=generic+bulk_memory --pkg-begin build_options /home/ci/actions-runner7/_work/zig/zig/build-debug/zig-local-cache/options/DnIZA_2_0w-HpqvELBNvknxl_3sFKNGudpRc-lyHGgtBMulNp2rHwXtJFlFMEShq --pkg-end --enable-cache 
2023-01-04T07:48:08.8980567Z error: the following build command failed with exit code 1:
2023-01-04T07:48:08.8981643Z /home/ci/actions-runner7/_work/zig/zig/build-debug/zig-local-cache/o/f2d0549f98984e84acc38602198cc57e/build /home/ci/actions-runner7/_work/zig/zig/build-debug/stage3-debug/bin/zig /home/ci/actions-runner7/_work/zig/zig /home/ci/actions-runner7/_work/zig/zig/build-debug/zig-local-cache /home/ci/actions-runner7/_work/zig/zig/build-debug/zig-global-cache update-zig1
2023-01-04T07:48:08.9034292Z ##[error]Process completed with exit code 1.
```

File: lib/std/debug.zig [hunk 2]
```diff
@@ -110,25 +110,31 @@ pub fn getSelfDebugInfo() !*DebugInfo {
...
+        if (process.hasEnvVarConstant("ZIG_DEBUG_COLOR")) {
             return .escape_codes;
-        } else if (native_os == .windows and file.isTty()) {
-            var info: windows.CONSOLE_SCREEN_BUFFER_INFO = undefined;
-            if (windows.kernel32.GetConsoleScreenBufferInfo(file.handle, &info) != windows.TRUE) {
-                // TODO: Should this return an error instead?
+        } else if (process.hasEnvVarConstant("NO_COLOR")) {
+            return .no_color;
+        } else {
+            if (file.supportsAnsiEscapeCodes()) {
```
Comment (Vexu) [inline lib/std/debug.zig:123]: Could you also replace the two `else { if`s in this function with `else if` when you handle the uses of `hasEnvVarConstant` in `main.zig`

**Comment by brendandburns** [comment]
@Vexu thanks for the patience as I got this working. I think this is now ready to merge.

**Pull request merged**